### DISCUSSION TITLE
Testing modm based software. Hosted application build.

### DISCUSSION THREAD

**Discussion by Tecnologic** [modm-io/modm#1244] [Q&A] [answered]
Hi all,

i'm starting with modm and i build my appication for a stm32 target. To test and develop my software i want to run tests on my host. I thought of having my application in the default ext/app/ tree and along side a test_app using code from the application for tests. Is there a way to have lbuild generate a hosted variant of modm to stub most of the modm references? The rest of the hardware singals needed i want to simulate in the hosted enviroment. Is this the right approach? How could a project.xml look like for the tests?

Thnaks

Alex

**Comment by salkinium** [comment, marked as answer]
Yes, modm defines [several `hosted-{platform}` targets here](https://github.com/modm-io/modm/tree/develop/tools/devices/hosted).

We have [some examples for these hosted targets](https://github.com/modm-io/modm/tree/develop/examples/linux) and we also run [the unit tests in the CI using these targets](https://github.com/modm-io/modm/blob/develop/test/config/hosted.xml) (the specific target option is added via command line in the [Makefile](https://github.com/modm-io/modm/blob/develop/test/Makefile#L29-L38) so we can support x86 and ARM64 using the same config file).

There are also some [mock ups of peripherals for the unit tests here](https://github.com/modm-io/modm/tree/develop/test/modm/mock).

**Comment by Tecnologic** [comment]
Thanks. I saw the tests but couldn't adapt it to my use. The examples are a great help.

**Reply by salkinium** [reply]
If you don't plan on running the same tests on hosted and *also* on a microcontroller, then I would recommend to use Google Test or Catch2 or similar. Our test harness is very limited and exists mostly just so that it works on small AVRs, but it's very barebone and lacks a lot of useful features.

**Reply by Tecnologic** [reply]
That's the plan. I want to use the cmake and gtest setup from a embedded Linux project for the modm project. And use the hosted variant only for the tests without the hardware.
